src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineViewModel_Navigation.cs

```diff
@@ -0,0 +1,52 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Collections;
+using Microsoft.CodeAnalysis.Navigation;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.DocumentOutline
+{
+    internal partial class DocumentOutlineViewModel
+    {
+        /// <summary>
+        /// A Queue of text spans to navigate to.
+        /// </summary>
+        private readonly AsyncBatchingWorkQueue<TextSpan> _navigationQueue;
+
+        internal void EnqueueNavigation(TextSpan textSpan)
+            => _navigationQueue.AddWork(textSpan, cancelExistingWork: true);
+
+        /// <summary>
+        /// Event handler that is called once we have finished navigating to a span.
+        /// </summary>
+        public event EventHandler? NavigationCompleted;
+
+        private async ValueTask NavigateToTextSpanAsync(ImmutableSegmentedList<TextSpan> textSpans, CancellationToken token)
+        {
+            var textSpan = textSpans.Last();
```

Comment:
so it's unclear to me why navigation is on a BG queue.  We should know all the locations of the items in the tree right, with appropriate ITrackingSpans for them.  So we should be able to effectively navigate immediately to them?  Or is that not the case.